# -- Merging dan GroupBy --


## Data Merging

Pandas menyediakan berbagai tool atau method yang memudahkan kita untuk melakukan penggabungan data dengan berbagai macam logik. method-method tersebut diantaranya seperti join(), merge(), concat(). kali ini kita kan mempelejari beberapa teknik tersebut.

Ada beberapa logika penggabungan data, diantaranya adalah :

1. Inner Join
2. Full Outer Join
3. Left Outer Join
4. Right Outer Join
5. dll.

#### Penggabungan Baris dan Column dengan Method CONCAT()

Concat adalah salah satu metode untuk melakukan penggabungan data berdasarkan sumbu tertentu. langsung saja kita lihat.

In [1]:
import pandas as pd

In [22]:
df1 = pd.DataFrame({
    'id_siswa' : ['A001','A002','A003','A004'],
    'Kelas' : ['12A','12B','12C','12D'],
    'Nilai' : [100,98,96,95],
    'Peringkat' : [1,2,3,4]
},index = [0,1,2,3])

In [3]:
df2 = pd.DataFrame({
    'id_siswa' : ['A005','A006','A007','A008'],
    'Kelas' : ['12E','12F','12G','12H'],
    'Nilai' : [90,88,86,85],
    'Peringkat' : [5,6,7,8]
}, index=[4,5,6,7])

In [4]:
df3 = pd.DataFrame({
    'id_siswa' : ['A009','A010','A011','A012'],
    'Kelas' : ['12I','12J','12K','12L'],
    'Nilai' : [80,75,73,70],
    'Peringkat' : [9,10,11,12]
}, index=[9,10,11,12])

### Concat ()

apabila kita ingin menggabungkan baris data ini, kita bisa menggunakan suatu method concat()

 pertama kita buat terlebih dahulu list dari data yang akan kita gabung, kemudian masukan list tersebut sebagai argument kedalam method concat()

In [24]:
# list data yang akan di gabung

list_dataframe = [df1,df2,df3]

# df1.loc["New"] =[1,2,3,43]
df1.loc["New"] = [11,22,33,44]
# df1.insert(loc=len(df1.columns),column="New",value=[1,2,3,4])
df1

id_siswa Kelas  Nilai  Peringkat
0       A001   12A    100          1
1       A002   12B     98          2
2       A003   12C     96          3
3       A004   12D     95          4
New       11    22     33         44

In [11]:
# gabungkan data dengan metode concat dari pandas

pd.concat(list_dataframe)

id_siswa Kelas  Nilai  Peringkat
0      A001   12A    100          1
1      A002   12B     98          2
2      A003   12C     96          3
3      A004   12D     95          4
4      A005   12E     90          5
5      A006   12F     88          6
6      A007   12G     86          7
7      A008   12H     85          8
9      A009   12I     80          9
10     A010   12J     75         10
11     A011   12K     73         11
12     A012   12L     70         12

Apabila kita ingin menggabungkan data tersebut berdasarkan column, kita mesti memberikan satu argument lagi kedalam method concat yaitu axis, dengan nilai 1. seperti ini

In [13]:
pd.concat(list_dataframe,axis=1)

id_siswa Kelas  Nilai  Peringkat id_siswa Kelas  Nilai  Peringkat id_siswa  \
0      A001   12A  100.0        1.0      NaN   NaN    NaN        NaN      NaN   
1      A002   12B   98.0        2.0      NaN   NaN    NaN        NaN      NaN   
2      A003   12C   96.0        3.0      NaN   NaN    NaN        NaN      NaN   
3      A004   12D   95.0        4.0      NaN   NaN    NaN        NaN      NaN   
4       NaN   NaN    NaN        NaN     A005   12E   90.0        5.0      NaN   
5       NaN   NaN    NaN        NaN     A006   12F   88.0        6.0      NaN   
6       NaN   NaN    NaN        NaN     A007   12G   86.0        7.0      NaN   
7       NaN   NaN    NaN        NaN     A008   12H   85.0        8.0      NaN   
9       NaN   NaN    NaN        NaN      NaN   NaN    NaN        NaN     A009   
10      NaN   NaN    NaN        NaN      NaN   NaN    NaN        NaN     A010   
11      NaN   NaN    NaN        NaN      NaN   NaN    NaN        NaN     A011   
12      NaN   NaN    NaN        NaN      NaN   NaN    NaN        NaN     A012   

   Kelas  Nilai  Peringkat  
0    NaN    NaN        NaN  
1    NaN    NaN        NaN  
2    NaN    NaN        NaN  
3    NaN    NaN        NaN  
4    NaN    NaN        NaN  
5    NaN    NaN        NaN  
6    NaN    NaN        NaN  
7    NaN    NaN        NaN  
9    12I   80.0        9.0  
10   12J   75.0       10.0  
11   12K   73.0       11.0  
12   12L   70.0       12.0

# Logika-Logika Penggabungan Dengan Metode merge

## Full Outer Join

1. Full Outer Join menghasilkan gabungan index bari dari setiap data yang di gabung.
2. Full Outer Join menggabungan pula column yang berbeda
3. Memberikan nilai null untuk nilai column yang tidak di miliki oleh suatu index tertentu.

In [24]:
dff = pd.DataFrame({
    'id_siswa' : ['A001','A002','A003','A004'],
    'Kelas' : ['12A','12B','12C','12D'],
    'Tinggi siswa (cm)' : [155,160,170,165]
}, index=[0,1,2,3])

In [30]:
dfff = pd.DataFrame({
    'id_siswa' : ['A001','A002','A003','A005'],
    'Nilai' : [100,98,96,95],
    'Peringkat' : [1,2,3,4]
},index=[4,5,6,7])

In [31]:
print('data pertama')
print("================")
print(dff)
print("")
print("=============================")
print("")
print("data kedua")
print("================")
print(dfff)

data pertama
  id_siswa Kelas  Tinggi siswa (cm)
0     A001   12A                155
1     A002   12B                160
2     A003   12C                170
3     A004   12D                165


data kedua
  id_siswa  Nilai  Peringkat
4     A001    100          1
5     A002     98          2
6     A003     96          3
7     A005     95          4


##### menggabungkan data ini dengan menggunakan method merge()

1. Untuk melakukan penggabungan dengan menggunakan column key sebagai referensi penggabungan, maka kita akan memberikan argument on dengan nilai 'id_siswa'
2. Untuk melakukan penggabungan outer join kita akan memberikan argument how dengan nilai 'outer'

In [21]:
# gabungkan data dengan logika outer join

In [32]:
pd.merge(dff,dfff, on="id_siswa", how='outer')

id_siswa Kelas  Tinggi siswa (cm)  Nilai  Peringkat
0     A001   12A              155.0  100.0        1.0
1     A002   12B              160.0   98.0        2.0
2     A003   12C              170.0   96.0        3.0
3     A004   12D              165.0    NaN        NaN
4     A005   NaN                NaN   95.0        4.0

## Inner Join 

Untuk melakukan inner join dengan metode merge(), kurang lebih sama seperti yang di lakukan pada outer join, kita hanya perlu mengubah argument argument how menjadi inner.

In [33]:
pd.merge(dff,dfff, on='id_siswa',how='inner')

id_siswa Kelas  Tinggi siswa (cm)  Nilai  Peringkat
0     A001   12A                155    100          1
1     A002   12B                160     98          2
2     A003   12C                170     96          3

## Left Join

Left Outer Join menghasilkan seluruh data dari data yang kiri ditambah data kanan yang memiliki nilai key yang sama dengan data dari kiri.

melakukan left outer join dengan data yang sama pada outer join, untuk melakukan left outer join kita hanya perlu memberikan nilai 'left' kepada argument how.

In [34]:
pd.merge(dff,dfff, on = 'id_siswa', how='left')

id_siswa Kelas  Tinggi siswa (cm)  Nilai  Peringkat
0     A001   12A                155  100.0        1.0
1     A002   12B                160   98.0        2.0
2     A003   12C                170   96.0        3.0
3     A004   12D                165    NaN        NaN

## Right Join 

melakukan operasi penggabungan right outer join dengan metode merge, kita hanya perlu memberikan nilai 'right' pada argument how.

In [35]:
pd.merge(dff,dfff, on="id_siswa", how='right')

id_siswa Kelas  Tinggi siswa (cm)  Nilai  Peringkat
0     A001   12A              155.0    100          1
1     A002   12B              160.0     98          2
2     A003   12C              170.0     96          3
3     A005   NaN                NaN     95          4

# Data GroupBy

Grouping adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. Untuk melakukan grouping di pandas kita menggunakan suatu method yang bernama **groupby()**. definisi abstraknya dari grouping adalah kita melakukan mapping suatu data terhadap suatu group.

ketika melakukan grouping ada beberapa proses yang terjadi secara berurutan, yaitu :

1. Splitting : Memisahkan data kedalam suatu group berdasarkan kriteria tertentu.
2. Applying : Melakukan suatu operasi terhadap sekumpulan data di group-group tersebut.
3. Combining : Menggabungkan data menjadi suatu struktur baru

In [1]:
import pandas as pd
import numpy as np
df_grup = pd.DataFrame([('bird','Falconiformes',389.0),
                       ('bird','Psittaciformes',24.0),
                       ('mammal','Carnivora',80.2),
                       ('mammal','Primates',np.nan),
                       ('mammal','Carnivora',58)],
                      index=['falcon','parrot','lion','monkey','leopard'],
                      columns=('class','order','max_speed'))

In [2]:
df_grup

class           order  max_speed
falcon     bird   Falconiformes      389.0
parrot     bird  Psittaciformes       24.0
lion     mammal       Carnivora       80.2
monkey   mammal        Primates        NaN
leopard  mammal       Carnivora       58.0

In [3]:
df_grup.groupby('class')

Tidak menghasilkan nilai apapun sebelum kita memberikan suatu operasi terhadap hasil dari setiap grouping tersebut. Prosess melakukan operasi terhadap sekumpulan data yang telah di groupkan adalah process applying. Kita bisa menggunakan berbagai jenis operasi aggregasi terhadap objek groupby tersebut. Ada operasi yang telah di sediakan sebagai operasi bawaan dari pandas seperti mean(), min(), max(), dan lain sebagainya.

In [39]:
df_grup.groupby('class').mean()

max_speed
class            
bird        206.5
mammal       69.1

Coba kalian perhatikan. kita telah berhasil menggunakan groupby dan memberikan hasil. jenis operasi yang di apply terhadap setiap group adalah rata-rata. dan hasilnya di combine menjadi struktur data baru. dengan begini kita menhasilkan suatu hasil analisis sederhana yaitu bahwa binatang dengan class bird memiliki kecepatan maksimal rata-rata sebesar 206.5 dan binatang class mamalia sebesar 69.1.

In [4]:
dd = df_grup.groupby(['class','order'])[['max_speed']].agg(['mean','sum'])

In [6]:
# dd.loc['bird'].loc['order']
# dd.iloc['bird'].order
# dd['order']['bird']


dd.iloc[1].iloc[0] 


dd



max_speed       
                           mean    sum
class  order                          
bird   Falconiformes      389.0  389.0
       Psittaciformes      24.0   24.0
mammal Carnivora           69.1  138.2
       Primates             NaN    0.0